### ISSUE TITLE
DDP RendezvousClosedError Exception

### ISSUE THREAD

**Issue by Mitchell-Chatterjee** [pytorch/pytorch#116974]
### 🐛 Describe the bug

I am running into the following RendezvousClosedError, for which I have attached the entire stack trace. I am training a model on multiple gpus, across multiple nodes of a High-Performance Cluster (Cedar -- Digital Alliance Canada).

However, the model runs for 2-3 epochs (approx 10 minutes) before it fails. This is consistent across at least 10 runs. I checked to see if any of the processes were exiting the script prematurely by printing "For gpu:x. Running epoch: 3" prior to the third epoch and it seems they are still running. The error message doesn't reveal much and I'm unsure where the problem resides.

Finally, I tried using tmux as some [other threads](https://github.com/pytorch/pytorch/issues/67538) alluded to, but that didn't seem to help. Granted I may not be using it correctly.

The only other piece of information that may be useful. Is that this seems to happen consistently around the 10 minute mark. There is some overhead, prior to the model beginning to train. I have attached my script and the stack trace below.

####################### Script #######################
#!/bin/bash
#SBATCH --account=name
#SBATCH --nodes=2
#SBATCH --gpus-per-node=v100l:2        # Request GPU "generic resources"
#SBATCH --tasks-per-node=2
#SBATCH --cpus-per-task=8   # Refer to cluster's documentation for the right CPU/GPU ratio
#SBATCH --mem=64000M           
#SBATCH --time=00:20:00     # DD-HH:MM:SS

module load python/3.10 cuda cudnn

#Prepare virtualenv
source /home/mitch/ENV/bin/activate

#Use same number of threads as cpus
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK

#Get the rendevous IP
nodes=( $( scontrol show hostnames $SLURM_JOB_NODELIST ) )
nodes_array=($nodes)
head_node=${nodes_array[0]}
hostname=$(srun --nodes=1 --ntasks=1 -w "$head_node" hostname)

echo Hostname: $hostname
export LOGLEVEL=INFO
export NCCL_DEBUG=INFO

#Move to correct directory
cd /scratch/mitch/git/Masters-Research/ || exit

#Start tmux
tmux new -d -s "$SLURM_JOB_ID"

#Run commands inside the Tmux session
tmux send-keys -t "$SLURM_JOB_ID" "echo hello world" Enter

#Torch run will naturally access the last checkpoint if interrupted
srun torchrun --nnodes=2 \
              --nproc_per_node=2 \
              --rdzv_id=$RANDOM \
              --rdzv_backend=c10d \
              --rdzv_endpoint=$hostname:29500 \
              -m src.patchtst_pretrain \
              --lr=0.0005 \
              --save_every=1 \
              --root_path=/scratch/mitch/datasets/500Hz/CODE_test \
              --num_workers=8 \
              --batch_size=64 \
              --n_epochs_pretrain=5 \
              --plot_every_n=1 \
              --distributed   \
              --model_name=pretrain_test_multi_node
####################### Stack Trace #######################
```
Due to MODULEPATH changes, the following have been reloaded:
  1) libfabric/1.10.1     2) openmpi/4.0.3     3) ucx/1.8.0

Hostname: cdr2596.int.cedar.computecanada.ca
master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
INFO:torch.distributed.launcher.api:Starting elastic_operator with launch configs:
  entrypoint       : src.patchtst_pretrain
  min_nodes        : 2
  max_nodes        : 2
  nproc_per_node   : 2
  run_id           : 1341
  rdzv_backend     : c10d
  rdzv_endpoint    : cdr2596.int.cedar.computecanada.ca:29500
  rdzv_configs     : {'timeout': '7200'}
  max_restarts     : 0
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

INFO:torch.distributed.launcher.api:Starting elastic_operator with launch configs:
  entrypoint       : src.patchtst_pretrain
  min_nodes        : 2
  max_nodes        : 2
  nproc_per_node   : 2
  run_id           : 1341
  rdzv_backend     : c10d
  rdzv_endpoint    : cdr2596.int.cedar.computecanada.ca:29500
  rdzv_configs     : {'timeout': '7200'}
  max_restarts     : 0
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

INFO:torch.distributed.elastic.agent.server.local_elastic_agent:log directory set to: /tmp/torchelastic_bs6hxenn/1341_lvsh_efc
INFO:torch.distributed.elastic.agent.server.local_elastic_agent:log directory set to: /tmp/torchelastic_j28azoe1/1341_7vnbcoa9
INFO:torch.distributed.elastic.agent.server.api:[default] starting workers for entrypoint: python
INFO:torch.distributed.elastic.agent.server.api:[default] starting workers for entrypoint: python
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous'ing worker group
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous'ing worker group
INFO:torch.distributed.launcher.api:Starting elastic_operator with launch configs:
  entrypoint       : src.patchtst_pretrain
  min_nodes        : 2
  max_nodes        : 2
  nproc_per_node   : 2
  run_id           : 1341
  rdzv_backend     : c10d
  rdzv_endpoint    : cdr2596.int.cedar.computecanada.ca:29500
  rdzv_configs     : {'timeout': '7200'}
  max_restarts     : 0
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
INFO:torch.distributed.launcher.api:Starting elastic_operator with launch configs:
  entrypoint       : src.patchtst_pretrain
  min_nodes        : 2
  max_nodes        : 2
  nproc_per_node   : 2
  run_id           : 1341
  rdzv_backend     : c10d
  rdzv_endpoint    : cdr2596.int.cedar.computecanada.ca:29500
  rdzv_configs     : {'timeout': '7200'}
  max_restarts     : 0
  monitor_interval : 5
  log_dir          : None
  metrics_cfg      : {}

INFO:torch.distributed.elastic.agent.server.local_elastic_agent:log directory set to: /tmp/torchelastic_lruxrkr4/1341_dhlv6jzo
INFO:torch.distributed.elastic.agent.server.api:[default] starting workers for entrypoint: python
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous'ing worker group
[W socket.cpp:426] [c10d] The server socket has failed to bind to [::]:29500 (errno: 98 - Address already in use).
[W socket.cpp:426] [c10d] The server socket has failed to bind to 0.0.0.0:29500 (errno: 98 - Address already in use).
[E socket.cpp:462] [c10d] The server socket has failed to listen on any local network address.
INFO:torch.distributed.elastic.agent.server.local_elastic_agent:log directory set to: /tmp/torchelastic_mgiaz502/1341_u5trhtze
INFO:torch.distributed.elastic.agent.server.api:[default] starting workers for entrypoint: python
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous'ing worker group
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous complete for workers. Result:
  restart_count=0
  master_addr=cdr2596.int.cedar.computecanada.ca
  master_port=49936
  group_rank=0
  group_world_size=2
  local_ranks=[0, 1]
  role_ranks=[0, 1]
  global_ranks=[0, 1]
  role_world_sizes=[4, 4]
  global_world_sizes=[4, 4]

INFO:torch.distributed.elastic.agent.server.api:[default] Starting worker group
INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous complete for workers. Result:
  restart_count=0
  master_addr=cdr2596.int.cedar.computecanada.ca
  master_port=49936
  group_rank=1
  group_world_size=2
  local_ranks=[0, 1]
  role_ranks=[2, 3]
  global_ranks=[2, 3]
  role_world_sizes=[4, 4]
  global_world_sizes=[4, 4]

INFO:torch.distributed.elastic.agent.server.api:[default] Starting worker group
INFO:torch.distributed.elastic.agent.server.local_elastic_agent:Environment variable 'TORCHELASTIC_ENABLE_FILE_TIMER' not found. Do not start FileTimerServer.
INFO:torch.distributed.elastic.multiprocessing:Setting worker0 reply file to: /tmp/torchelastic_lruxrkr4/1341_dhlv6jzo/attempt_0/0/error.json
INFO:torch.distributed.elastic.multiprocessing:Setting worker1 reply file to: /tmp/torchelastic_lruxrkr4/1341_dhlv6jzo/attempt_0/1/error.json
INFO:torch.distributed.elastic.agent.server.local_elastic_agent:Environment variable 'TORCHELASTIC_ENABLE_FILE_TIMER' not found. Do not start FileTimerServer.
INFO:torch.distributed.elastic.multiprocessing:Setting worker0 reply file to: /tmp/torchelastic_bs6hxenn/1341_lvsh_efc/attempt_0/0/error.json
INFO:torch.distributed.elastic.multiprocessing:Setting worker1 reply file to: /tmp/torchelastic_bs6hxenn/1341_lvsh_efc/attempt_0/1/error.json


Executing program in Distributed Data Parallel mode
Number of GPUs available on each node: 2

args: Namespace(dset_pretrain='CODE_Unannotated', root_path='/scratch/mitch/datasets/500Hz/CODE_test', top_n_classes=None, min_class_size=50, superclass_classification=False, distributed=True, data_sub_path='training', context_points=5000, target_points=100, batch_size=64, num_workers=8, scaler='standard', features='M', patch_len=20, stride=20, revin=1, learning_rate=None, n_layers=3, n_heads=16, d_model=256, d_ff=512, dropout=0.2, head_type=<Mode.PRETRAIN: 1>, head_dropout=0.2, mask_ratio=0.4, root_path_noise='/home/student/Datasets/noise', noise_masking=False, save_every=1, plot_every_n=1, n_epochs_pretrain=5, lr=0.0005, pretrained_model_id=1, model_type='based_model', model_name='pretrain_test_multi_node_3_', save_path='results/pre-train/saved_models/CODE_Unannotated/masked_patchtst/based_model/pretrain_test_multi_node_3_/', dset='CODE_Unannotated')

Namespace(dset_pretrain='CODE_Unannotated', root_path='/scratch/mitch/datasets/500Hz/CODE_test', top_n_classes=None, min_class_size=50, superclass_classification=False, distributed=True, data_sub_path='training', context_points=5000, target_points=100, batch_size=64, num_workers=8, scaler='standard', features='M', patch_len=20, stride=20, revin=1, learning_rate=None, n_layers=3, n_heads=16, d_model=256, d_ff=512, dropout=0.2, head_type=<Mode.PRETRAIN: 1>, head_dropout=0.2, mask_ratio=0.4, root_path_noise='/home/student/Datasets/noise', noise_masking=False, save_every=1, plot_every_n=1, n_epochs_pretrain=5, lr=0.0005, pretrained_model_id=1, model_type='based_model', model_name='pretrain_test_multi_node_3_', save_path='results/pre-train/saved_models/CODE_Unannotated/masked_patchtst/based_model/pretrain_test_multi_node_3_/', dset='CODE_Unannotated')
number of patches: 250
number of model params 1655828
cdr2596:222141:222141 [0] NCCL INFO Bootstrap : Using ib0:172.19.146.33<0>
cdr2596:222141:222141 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
cdr2596:222141:222141 [0] NCCL INFO NET/IB : Using [0]hfi1_0:1/IB ; OOB ib0:172.19.146.33<0>
cdr2596:222141:222141 [0] NCCL INFO Using network IB
NCCL version 2.12.12+cuda11.7
cdr2596:222142:222142 [1] NCCL INFO Bootstrap : Using ib0:172.19.146.33<0>
cdr2596:222142:222142 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
cdr2596:222142:222142 [1] NCCL INFO NET/IB : Using [0]hfi1_0:1/IB ; OOB ib0:172.19.146.33<0>
cdr2596:222142:222142 [1] NCCL INFO Using network IB
cdr2598:80003:80003 [0] NCCL INFO Bootstrap : Using ib0:172.19.146.35<0>
cdr2598:80004:80004 [1] NCCL INFO Bootstrap : Using ib0:172.19.146.35<0>
cdr2598:80004:80004 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
cdr2598:80003:80003 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
cdr2598:80003:80003 [0] NCCL INFO NET/IB : Using [0]hfi1_0:1/IB ; OOB ib0:172.19.146.35<0>
cdr2598:80004:80004 [1] NCCL INFO NET/IB : Using [0]hfi1_0:1/IB ; OOB ib0:172.19.146.35<0>
cdr2598:80004:80004 [1] NCCL INFO Using network IB
cdr2598:80003:80003 [0] NCCL INFO Using network IB
cdr2598:80004:80116 [1] NCCL INFO Setting affinity for GPU 1 to aaa028
cdr2598:80003:80115 [0] NCCL INFO Setting affinity for GPU 0 to 155405
cdr2596:222142:222213 [1] NCCL INFO Setting affinity for GPU 1 to 541415
cdr2596:222141:222211 [0] NCCL INFO Setting affinity for GPU 0 to 541415
cdr2596:222142:222213 [1] NCCL INFO Trees [0] -1/-1/-1->1->0 [1] -1/-1/-1->1->0
cdr2596:222141:222211 [0] NCCL INFO Channel 00/02 :    0   1   2   3
cdr2596:222141:222211 [0] NCCL INFO Channel 01/02 :    0   1   2   3
cdr2596:222141:222211 [0] NCCL INFO Trees [0] 1/2/-1->0->-1 [1] 1/-1/-1->0->2
cdr2598:80003:80115 [0] NCCL INFO Trees [0] 3/-1/-1->2->0 [1] 3/0/-1->2->-1
cdr2598:80004:80116 [1] NCCL INFO Trees [0] -1/-1/-1->3->2 [1] -1/-1/-1->3->2
cdr2598:80003:80115 [0] NCCL INFO Channel 00/0 : 1[3b000] -> 2[18000] [receive] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Channel 00/0 : 3[86000] -> 0[18000] [receive] via NET/IB/0
cdr2598:80004:80116 [1] NCCL INFO Channel 00/0 : 3[86000] -> 0[18000] [send] via NET/IB/0
cdr2596:222142:222213 [1] NCCL INFO Channel 00/0 : 1[3b000] -> 2[18000] [send] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 01/0 : 1[3b000] -> 2[18000] [receive] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 00 : 2[18000] -> 3[86000] via P2P/IPC
cdr2596:222141:222211 [0] NCCL INFO Channel 01/0 : 3[86000] -> 0[18000] [receive] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Channel 00 : 0[18000] -> 1[3b000] via P2P/IPC
cdr2598:80003:80115 [0] NCCL INFO Channel 01 : 2[18000] -> 3[86000] via P2P/IPC
cdr2596:222141:222211 [0] NCCL INFO Channel 01 : 0[18000] -> 1[3b000] via P2P/IPC
cdr2598:80004:80116 [1] NCCL INFO Channel 01/0 : 3[86000] -> 0[18000] [send] via NET/IB/0
cdr2596:222142:222213 [1] NCCL INFO Channel 01/0 : 1[3b000] -> 2[18000] [send] via NET/IB/0
cdr2598:80004:80116 [1] NCCL INFO Connected all rings
cdr2598:80004:80116 [1] NCCL INFO Channel 00 : 3[86000] -> 2[18000] via P2P/IPC
cdr2598:80004:80116 [1] NCCL INFO Channel 01 : 3[86000] -> 2[18000] via P2P/IPC
cdr2596:222141:222211 [0] NCCL INFO Connected all rings
cdr2596:222142:222213 [1] NCCL INFO Connected all rings
cdr2596:222142:222213 [1] NCCL INFO Channel 00 : 1[3b000] -> 0[18000] via P2P/IPC
cdr2596:222142:222213 [1] NCCL INFO Channel 01 : 1[3b000] -> 0[18000] via P2P/IPC
cdr2598:80003:80115 [0] NCCL INFO Connected all rings
cdr2596:222141:222211 [0] NCCL INFO Channel 00/0 : 2[18000] -> 0[18000] [receive] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 00/0 : 0[18000] -> 2[18000] [receive] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Channel 01/0 : 2[18000] -> 0[18000] [receive] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 01/0 : 0[18000] -> 2[18000] [receive] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Channel 00/0 : 0[18000] -> 2[18000] [send] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 00/0 : 2[18000] -> 0[18000] [send] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Channel 01/0 : 0[18000] -> 2[18000] [send] via NET/IB/0
cdr2598:80003:80115 [0] NCCL INFO Channel 01/0 : 2[18000] -> 0[18000] [send] via NET/IB/0
cdr2596:222141:222211 [0] NCCL INFO Connected all trees
cdr2596:222141:222211 [0] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
cdr2596:222141:222211 [0] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
cdr2596:222142:222213 [1] NCCL INFO Connected all trees
cdr2596:222142:222213 [1] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
cdr2596:222142:222213 [1] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
cdr2598:80003:80115 [0] NCCL INFO Connected all trees
cdr2598:80003:80115 [0] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
cdr2598:80003:80115 [0] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
cdr2598:80004:80116 [1] NCCL INFO Connected all trees
cdr2598:80004:80116 [1] NCCL INFO threadThresholds 8/8/64 | 32/8/64 | 8/8/512
cdr2598:80004:80116 [1] NCCL INFO 2 coll channels, 2 p2p channels, 2 p2p channels per peer
cdr2596:222142:222213 [1] NCCL INFO comm 0x2b847c0090d0 rank 1 nranks 4 cudaDev 1 busId 3b000 - Init COMPLETE
cdr2596:222141:222211 [0] NCCL INFO comm 0x2adf000090d0 rank 0 nranks 4 cudaDev 0 busId 18000 - Init COMPLETE
cdr2598:80004:80116 [1] NCCL INFO comm 0x2aff340090d0 rank 3 nranks 4 cudaDev 1 busId 86000 - Init COMPLETE
cdr2598:80003:80115 [0] NCCL INFO comm 0x2b69500090d0 rank 2 nranks 4 cudaDev 0 busId 18000 - Init COMPLETE
cdr2596:222141:222141 [0] NCCL INFO Launch mode Parallel
Device: cuda:3
Device: cuda:0Device: cuda:1

Device: cuda:2
For gpu:3. Running epoch: 0
For gpu:1. Running epoch: 0
For gpu:2. Running epoch: 0

Batchsize: 64 | Steps: 277

                   gpu_id                    epoch               train_loss               valid_loss                     time
For gpu:0. Running epoch: 0
                        1                        0                 0.993345                 0.999450                    02:40
For gpu:1. Running epoch: 1
                        2                        0                 0.998787                 0.990844                    02:40
For gpu:2. Running epoch: 1
                        3                        0                 0.998005                 0.984352                    02:40
For gpu:3. Running epoch: 1
                        0                        0                 0.992754                 0.979189                    02:33
Saved model and snapshot at epoch 0
Better model found at epoch 0 with valid_loss value: 0.9791890885879571.
For gpu:0. Running epoch: 1
                        1                        1                 0.967605                 0.942914                    02:32
For gpu:1. Running epoch: 2
                        3                        1                 0.965737                 0.940590                    02:32
For gpu:3. Running epoch: 2
                        2                        1                 0.972108                 0.947598                    02:32
For gpu:2. Running epoch: 2
                        0                        1                 0.965889                 0.946714                    02:28
Saved model and snapshot at epoch 1
Better model found at epoch 1 with valid_loss value: 0.9467136117635856.
For gpu:0. Running epoch: 2
                        2                        2                 0.944043                 0.922976                    02:31
For gpu:2. Running epoch: 3
                        3                        2                 0.943278                 0.933801                    02:31
For gpu:3. Running epoch: 3
                        1                        2                 0.940759                 0.925462                    02:32
For gpu:1. Running epoch: 3
                        0                        2                 0.939825                 0.939667                    02:27
Saved model and snapshot at epoch 2
Better model found at epoch 2 with valid_loss value: 0.9396670735263776.
For gpu:0. Running epoch: 3
Traceback (most recent call last):
  File "/home/mitch/ENV/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 794, in main
    run(args)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 241, in launch_agent
    result = agent.run()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 723, in run
    result = self._invoke_run(role)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 858, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 692, in _initialize_workers
    self._rendezvous(worker_group)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 546, in _rendezvous
    store, group_rank, group_world_size = spec.rdzv_handler.next_rendezvous()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 1028, in next_rendezvous
    self._op_executor.run(join_op, deadline)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 638, in run
    raise RendezvousTimeoutError()
torch.distributed.elastic.rendezvous.api.RendezvousTimeoutError
Traceback (most recent call last):
  File "/home/mitch/ENV/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 794, in main
    run(args)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 241, in launch_agent
    result = agent.run()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 723, in run
    result = self._invoke_run(role)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 858, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 692, in _initialize_workers
    self._rendezvous(worker_group)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 546, in _rendezvous
    store, group_rank, group_world_size = spec.rdzv_handler.next_rendezvous()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 1028, in next_rendezvous
    self._op_executor.run(join_op, deadline)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 635, in run
    raise RendezvousClosedError()
torch.distributed.elastic.rendezvous.api.RendezvousClosedError
srun: error: cdr2596: task 1: Exited with exit code 1
srun: Terminating StepId=21941655.1
srun: error: cdr2598: task 2: Exited with exit code 1
slurmstepd: error: *** STEP 21941655.1 ON cdr2596 CANCELLED AT 2024-01-05T14:27:50 ***
WARNING:torch.distributed.elastic.agent.server.api:Received 15 death signal, shutting down workers
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 222141 closing signal SIGTERM
WARNING:torch.distributed.elastic.agent.server.api:Received 15 death signal, shutting down workers
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 80003 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 222142 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 80004 closing signal SIGTERM
Traceback (most recent call last):
  File "/home/mitch/ENV/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 794, in main
    run(args)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 241, in launch_agent
    result = agent.run()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 723, in run
    result = self._invoke_run(role)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 864, in _invoke_run
    time.sleep(monitor_interval)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 62, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 79920 got signal: 15
Traceback (most recent call last):
  File "/home/mitch/ENV/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 794, in main
    run(args)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 241, in launch_agent
    result = agent.run()
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 129, in wrapper
    result = f(*args, **kwargs)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 723, in run
    result = self._invoke_run(role)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 864, in _invoke_run
    time.sleep(monitor_interval)
  File "/home/mitch/ENV/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/api.py", line 62, in _terminate_process_handler
    raise SignalException(f"Process {os.getpid()} got signal: {sigval}", sigval=sigval)
torch.distributed.elastic.multiprocessing.api.SignalException: Process 221925 got signal: 15
srun: error: cdr2598: task 3: Exited with exit code 1
srun: error: cdr2596: task 0: Exited with exit code 1
```
### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Gentoo Base System release 2.6 (x86_64)
GCC version: (GCC) 9.3.0
Clang version: Could not collect
CMake version: version 3.27.7
Libc version: glibc-2.30

Python version: 3.10.2 (main, Feb  4 2022, 19:10:35) [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.90.1.el7.x86_64-x86_64-with-glibc2.30
CPU:
Architecture:        x86_64

Versions of relevant libraries:
[pip3] numpy==1.25.2+computecanada
[pip3] torch==2.0.1+computecanada
[pip3] torcheval==0.0.7+computecanada

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

**Comment by wconstab** [comment]
@kurman could you take a look?

@Mitchell-Chatterjee I'm not sure if this stacktrace is the right one.  It's possible that the real error happened on a different process, and when that process closed it caused this rendezvous-failed error on a second process.

Do you have any alternative stacktraces that might be informative?

**Comment by Mitchell-Chatterjee** [comment]
@wconstab As far as I understand, slurm prints all output to the stacktrace I have provided. If there is another stacktrace I can try to dig it up, but as far as I know this is the only one. Do you mean that each process may have its own stacktrace?

That's also what I was thinking and it's certainly possible. Which is why I also included the print statement "For gpu:x. Running epoch: 3" to ensure the process was still running. Otherwise, there doesn't seem to be an error output related to my program as far as I can tell. Though it's definitely possible the error is on my end. I'll continue to look through my code.

For a bit more context. On a single node, with 4 GPUs. There is no issue with the program.

**Comment by kurman** [comment]
@Mitchell-Chatterjee I noticed following warning log message:
~~~
[W socket.cpp:426] [c10d] The server socket has failed to bind to [::]:29500 (errno: 98 - Address already in use).
[W socket.cpp:426] [c10d] The server socket has failed to bind to 0.0.0.0:29500 (errno: 98 - Address already in use).
[E socket.cpp:462] [c10d] The server socket has failed to listen on any local network address.
~~~

I suspect you have other jobs using the same node and same port. 10 mins most likely is related to `join` timeout - default value is 600 seconds. Just to validate, can you try updating `--rdzv_endpoint=$hostname:29500` parameter with some other available port?

**Comment by Mitchell-Chatterjee** [comment]
@kurman @wconstab 

Hi! Just wanted to get back to you and let you know that I found an alternative solution. I'm not entirely sure what the issue was with the torchrun code. I tried following your steps above by changing the port but that didn't seem to fix it.

However, I did find [this wiki](https://docs.alliancecan.ca/wiki/PyTorch/en#PyTorch_with_multiple_GPUs). Which I will link here if anyone else runs into the same issue.

It's likely some vagary of the cluster itself, so I should have been following their cluster specific documentation from the start.

Thanks again for looking into it!
